sonar-scm-svn-plugin/src/main/java/org/sonar/plugins/scm/svn/ChangedLinesComputer.java

```diff
@@ -0,0 +1,163 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.scm.svn;
+
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class ChangedLinesComputer {
+
+  private final Tracker tracker;
+
+  private final OutputStream receiver = new OutputStream() {
+    StringBuilder sb = new StringBuilder();
+
+    @Override
+    public void write(int b) {
+      sb.append((char) b);
+      if (b == '\n') {
+        tracker.parseLine(sb.toString());
+        sb.setLength(0);
+      }
+    }
+  };
+
+  ChangedLinesComputer(Set<Path> included) {
+    this.tracker = new Tracker(included);
+  }
+
+  OutputStream receiver() {
+    return receiver;
+  }
+
+  /**
+   * From a stream of svn-style unified diff lines,
+   * compute the line numbers that should be considered changed.
+   *
+   * Example input:
+   * <pre>
+   * Index: path/to/file
+   * ===================================================================
+   * --- lao 2002-02-21 23:30:39.942229878 -0800
+   * +++ tzu 2002-02-21 23:30:50.442260588 -0800
+   * @@ -1,7 +1,6 @@
+   * -The Way that can be told of is not the eternal Way;
+   * -The name that can be named is not the eternal name.
+   *  The Nameless is the origin of Heaven and Earth;
+   * -The Named is the mother of all things.
+   * +The named is the mother of all things.
+   * +
+   *  Therefore let there always be non-being,
+   *    so we may see their subtlety,
+   *  And let there always be being,
+   * @@ -9,3 +8,6 @@
+   *  The two are the same,
+   *  But after they are produced,
+   *    they have different names.
+   * +They both may be called deep and profound.
+   * +Deeper and more profound,
+   * +The door of all subtleties!
+   * </pre>
+   *
+   * See also: http://www.gnu.org/software/diffutils/manual/html_node/Example-Unified.html#Example-Unified
+   */
+  Map<Path, Set<Integer>> changedLines() {
+    return tracker.changedLines();
+  }
+
+  private static class Tracker {
+
+    private static final Pattern START_LINE_IN_TARGET = Pattern.compile(" \\+(\\d+)");
+    private static final String ENTRY_START_PREFIX = "Index: ";
+
+    private final Map<Path, Set<Integer>> changedLines = new HashMap<>();
+    private final Set<Path> included;
+
+    private int lineNumInTarget;
+    private Path currentPath = null;
+    private int skipCount = 0;
+
+    Tracker(Set<Path> included) {
+      this.included = included;
+    }
+
+    private void parseContent(String line) {
+      char firstChar = line.charAt(0);
+      if (firstChar == ' ') {
+        lineNumInTarget++;
+      } else if (firstChar == '+') {
+        changedLines.get(currentPath).add(lineNumInTarget);
```

Comment:
I think this could be `changedLines.computeIfAbsent(currentPath, HashMap::new).add(lineNumInTarget)` and we could avoid the `put` and the `removeCurrentChangesetIfEmpty`.